### PR TITLE
Permit to specify the banking customer length (up to 10 digits)

### PR DESCRIPTION
**Pull request by sdespont**
Marc,

I am facing a limitation: the banking customer ID I need to use has 7 digits and not 6 as usual. The current stable version expects the banking customer ID has exactly 6 digits. I have extended the class to use a banking customer ID up to 10 digits.

Also, I have added some verifications regarding the complete reference number when using with a banking customer ID which must have exactly 26 digits (without the check digit) and must be numerical.

Please, review these changes and tell me what you think about this suggestion.

Sébastien

### CODE REVIEW

**Comment by ravage84** [comment]
Hi @sdespont 
I'm very sorry, but I did not have enough time to look into it, yet.

**Comment by sdespont** [comment]
No problem, take your time. I am using my private fork until your review theses changes. In all cases, some banks started to expect at least 7 digits.

**Comment by ravage84** [comment]
> ... some banks started to expect at least 7 digits.

Can you add some links/resources about that?

**Comment by ravage84** [comment]
@sdespont I'm so sorry, that I did not catch up on this. But I'm swamped in work in the past months. And the other work in not related to this library, so my focus is totally somewhere else.

Could you please explain why you need a longer banking number? Do you have any official documentation on this? If so, we can base this PR on this documenttaion and even add a link to it.

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,11 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
 
 ### Fixed
+## [0.13.0]
```
Comment (ravage84) [inline CHANGELOG.md]: You can just keep it under "unreleased". When I release a version, I'm doing the change log update etc.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,11 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+- Add parameter to specify the length of the banking customer ID (up to 10) - 20161225
```
Comment (ravage84) [inline CHANGELOG.md]: No need to append the date at the end.

Comment (ravage84) [reply CHANGELOG.md]: I'd rather add the issue number ``#27`` at the end. That's much more helpful.

File: src/OrangePaymentSlipData.php [hunk 3]
```diff
@@ -54,6 +55,13 @@ class OrangePaymentSlipData extends PaymentSlipData
      * @var string
      */
     protected $bankingCustomerId = '';
+    
+    /**
+     * The banking customer ID expected length
+     *
+     * @var integer
+     */
+    protected $bankingCustomerIdLength = 0;
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php:69]: Setting it to zero seems to be a bit silly. Either unset or a reasonable default.

File: src/OrangePaymentSlipData.php [hunk 4]
```diff
@@ -100,6 +108,7 @@ public function setWithBankingCustomerId($withBankingCustomerId = true)
 
         if ($withBankingCustomerId === false) {
             $this->bankingCustomerId = '';
+            $this->bankingCustomerIdLength = 0;
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php:116]: Setting it to zero seems to be a bit silly. Either unset or a reasonable default.

File: src/OrangePaymentSlipData.php [hunk 5]
```diff
@@ -121,12 +130,19 @@ public function getWithBankingCustomerId()
      * @param string $referenceNumber The reference number.
      * @return $this The current instance for a fluent interface.
      * @throws DisabledDataException If the data is disabled.
+     * @throws PaymentSlipException If the length exceeds 26 when adding banking customer length.
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``ReferenceNumberException``.

File: src/OrangePaymentSlipData.php [hunk 6]
```diff
@@ -121,12 +130,19 @@ public function getWithBankingCustomerId()
...
      */
     public function setReferenceNumber($referenceNumber)
     {
         if (!$this->getWithReferenceNumber()) {
             throw new DisabledDataException('reference number');
         }
+                
+        if (strlen($referenceNumber) + $this->bankingCustomerIdLength > 26) {
+            throw new PaymentSlipException('Reference number will exceed the maximum '
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``ReferenceNumberException``.

File: src/OrangePaymentSlipData.php [hunk 7]
```diff
@@ -121,12 +130,19 @@ public function getWithBankingCustomerId()
...
      * @throws DisabledDataException If the data is disabled.
+     * @throws PaymentSlipException If the length exceeds 26 when adding banking customer length.
      */
     public function setReferenceNumber($referenceNumber)
     {
         if (!$this->getWithReferenceNumber()) {
             throw new DisabledDataException('reference number');
         }
+                
+        if (strlen($referenceNumber) + $this->bankingCustomerIdLength > 26) {
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: We should not use magic numbers, I prefer to have a constant for that, as this number cannot be changed. E.g. ``static::REFNR_MAX_LENGTH``

File: src/OrangePaymentSlipData.php [hunk 8]
```diff
@@ -121,12 +130,19 @@ public function getWithBankingCustomerId()
...
+            throw new PaymentSlipException('Reference number will exceed the maximum '
+                    . 'of 26 digits when banking customer will be prepended');
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: The error message could use the new constant.

File: src/OrangePaymentSlipData.php [hunk 9]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
      * @param string $bankingCustomerId The banking customer ID.
      * @return $this The current instance for a fluent interface.
      * @throws DisabledDataException If the data is disabled.
+     * @throws PaymentSlipException If the length is greater than 10 digits
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``BankingCustomerIdException``.

File: src/OrangePaymentSlipData.php [hunk 10]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
...
-    public function setBankingCustomerId($bankingCustomerId)
+    public function setBankingCustomerId($bankingCustomerId, $bankingCustomerIdLength = 6)
     {
         if (!$this->getWithBankingCustomerId()) {
             throw new DisabledDataException('banking customer ID');
         }
-        // TODO check length (exactly 6)
+        
+        if ($bankingCustomerIdLength > 10) {
+            throw new PaymentSlipException('Banking cutomer ID must not be longer than 10 digits');
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``BankingCustomerIdException``.

File: src/OrangePaymentSlipData.php [hunk 11]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
...
      */
-    public function setBankingCustomerId($bankingCustomerId)
+    public function setBankingCustomerId($bankingCustomerId, $bankingCustomerIdLength = 6)
     {
         if (!$this->getWithBankingCustomerId()) {
             throw new DisabledDataException('banking customer ID');
         }
-        // TODO check length (exactly 6)
+        
+        if ($bankingCustomerIdLength > 10) {
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: We should not use magic numbers, I prefer to have a constant for that, as this number cannot be changed. E.g. ``static::BANKING_CUST_ID_MAX_LENGTH``

File: src/OrangePaymentSlipData.php (hunk 10 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php]: This should use the new constant, then.

File: src/OrangePaymentSlipData.php [hunk 12]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
...
+            throw new PaymentSlipException('Banking cutomer ID must not be longer than 10 digits');
+        }
+        
+        if (strlen($bankingCustomerId) > $bankingCustomerIdLength) {
+            throw new PaymentSlipException('Banking cutomer ID must not be longer than the specified length');
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: State the specific length in the message. Helps debugging.

Comment (ravage84) [reply src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``BankingCustomerIdException``.

File: src/OrangePaymentSlipData.php [hunk 13]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
...
+        }
+        
+        if (strlen($this->referenceNumber) + $bankingCustomerIdLength > 26) {
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: These lines are duplicated. Please extract them into a protected method. Generalize where possible, e.g. through parameters.

File: src/OrangePaymentSlipData.php [hunk 14]
```diff
@@ -153,20 +169,35 @@ public function getReferenceNumber()
...
+        
+        $this->bankingCustomerIdLength = $bankingCustomerIdLength;
         $this->bankingCustomerId = $bankingCustomerId;
 
         return $this;
     }
 
     /**
-     * Get the banking customer ID
+     * Get the banking customer ID padded with 0
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: "with zeroes"

File: src/OrangePaymentSlipData.php [hunk 15]
```diff
@@ -176,7 +207,8 @@ public function getBankingCustomerId()
         if (!$this->getWithBankingCustomerId()) {
             throw new DisabledDataException('banking customer ID');
         }
-        return $this->bankingCustomerId;
+        
+        return str_pad($this->bankingCustomerId, $this->bankingCustomerIdLength, '0', STR_PAD_LEFT);
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php:219]: Not sure we always want that... Could we extract that into a separate PR?

File: src/OrangePaymentSlipData.php [hunk 16]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
...
+            
+            if (!is_numeric($completeReferenceNumber)) {
+                throw new PaymentSlipException('Reference number must be a numerical value');
+            }
         } elseif ($fillZeros) {
             // Get reference number and fill with zeros
             $completeReferenceNumber = str_pad($referenceNumber, 26, '0', STR_PAD_LEFT);
         }
-
+
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: Unecessary whitespace

File: src/OrangePaymentSlipData.php [hunk 17]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
...
         } elseif ($fillZeros) {
             // Get reference number and fill with zeros
             $completeReferenceNumber = str_pad($referenceNumber, 26, '0', STR_PAD_LEFT);
         }
-
+        
         // Add check digit
         $completeReferenceNumber = $this->appendCheckDigit($completeReferenceNumber, $notForPayment);
-
+
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: Unecessary whitespace

File: src/OrangePaymentSlipData.php [hunk 18]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
             $completeReferenceNumber = str_pad($referenceNumber, 26, 'X', STR_PAD_LEFT);
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: We should not use magic numbers, I prefer to have a constant for that, as this number cannot be changed. E.g. ``static::REFNR_MAX_LENGTH``.
And  yeah, that's my bad 😼

File: src/OrangePaymentSlipData.php [hunk 19]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
...
         } elseif ($this->getWithBankingCustomerId()) {
             // Get reference number and fill with zeros
-            $referenceNumber = str_pad($referenceNumber, 20, '0', STR_PAD_LEFT);
+            $padLength = 26 - $this->bankingCustomerIdLength;
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: We should not use magic numbers, I prefer to have a constant for that, as this number cannot be changed. E.g. ``static::REFNR_MAX_LENGTH``

File: src/OrangePaymentSlipData.php [hunk 20]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
...
+            $referenceNumber = str_pad($referenceNumber, $padLength, '0', STR_PAD_LEFT);
             // Prepend banking customer identification code
             $completeReferenceNumber = $this->getBankingCustomerId() . $referenceNumber;
+            
+            if (strlen($completeReferenceNumber) !== 26) {
+                throw new PaymentSlipException('Reference number must have 26 digits');
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``ReferenceNumberException``.

File: src/OrangePaymentSlipData.php [hunk 21]
```diff
@@ -220,17 +253,26 @@ public function getCompleteReferenceNumber($formatted = true, $fillZeros = true)
...
+            }
+            
+            if (!is_numeric($completeReferenceNumber)) {
+                throw new PaymentSlipException('Reference number must be a numerical value');
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``ReferenceNumberException``.

File: src/OrangePaymentSlipData.php [hunk 22]
```diff
@@ -209,6 +241,7 @@ public function setNotForPayment($notForPayment = true)
      * @param bool $formatted Should the returned reference be formatted in blocks of five (for better readability).
      * @param bool $fillZeros Fill up with leading zeros, only applies to the case where no banking customer ID is used.
      * @return string The complete (with/without bank customer ID), formatted reference number with check digit
+     * @throws PaymentSlipException if wrong length or is not numerical
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: A specific exception would make more sense, e.g. ``ReferenceNumberException``.

File: src/OrangePaymentSlipData.php (hunk 21 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php]: Often, I echo out the variable which lead to the exception, e.g. ``Reference number "%s" must be a numerical value``

**Review by ravage84 — CHANGES_REQUESTED** [review]
Many thanks again for this PR. See my comments and change requests.

For the future, please try to split up the work in smaller changes, e.g.  bank customer ID padding as separte PR.

**Comment by sdespont** [comment]
Hi @ravage84 ,

Some months ago, we had to use banking accounts with banking numbers up to 7 then 8 length for the Raiffeisen.

I asked the bank to receive official documentation about this, but there is no way that the bank gives details about their way to manage internal banking account.

Our client sent us the sample document given by the bank, it is all I have to share with you : [https://ibb.co/gD0Z7a](https://ibb.co/gD0Z7a)

**Comment by ravage84** [comment]
@sdespont thanks for your answer.

I had a further look into it. Though I did not find anything very useful, here are my findings:

https://www.raiffeisen.ch/content/dam/www/rch/pdf/firmenkunden/zahlungsverkehr-und-konten/de/Produktblatt-Zahlungsbeleg-ESR.pdf

> ESR-Identifikationsnummer des Kunden
> Die Nummer identifiziert Sie als Rechnungssteller und muss bei der Konfiguration
Ihrer Buchhaltungssoftware erfasst werden. Die Länge richtet sich nach den Anforderungen
> Ihrer Software und ist in der Regel 6-stellig.

__Normally__ 6 digits.

https://de.wikipedia.org/wiki/Kontonummer#Schweiz

> Kontonummern in der Schweiz haben bis zu 16 Stellen.[

Swiss bank account nubmers can be up to 16 digits

My personal bank account number is 641652-70-1, shortened 641652701, 9 digits long.

I did find another example of seven digits [here](http://wiki.4simplicity.ch/rechnungen/esr---einzahlungsschein).

https://www.zkb.ch/media/dok/efinance/vesr-handbuch-fk.pdf

> An Stelle der eigentlichen Kontonummer (11-stellig) wird ein 6-stellige Kundenidentifikationsnummer eingesetzt.

Though a normal banking account number is 11 digits, it will be repplaced with a 6 digit banking customer ID on a payment slip.

I'm going to write a mail to the Raiffaisen Bank. Let's see what they say

**Comment by sdespont** [comment]
Hi @ravage84 ,

Caution: do not confuse the banking account number with the banking customer identifier. 

In summary, the banking account number is the account of the bank in PostFinance bank. **All** the ESR are payed to PostFinance, then PostFinance send the money to the bank (BCV, Credit Suisse, UBS, ...).

So for each ESR, the bank (BCV, Credit Suisse, UBS, ...) receives the money from PostFinance. Then, the bank looks into the reference number and match the final customer account with the banking customer identifier (usually the 6 first digits). BUT, it is up to the bank to choose how they want to match the customer banking account and 6 digits is a strong limitation if you have a lot of clients (futhermore, the identifier is never reused for another client if the customer close its account). 

It is why a bank can decide to use 7 or 8 digits intead of 6 to match the customer banking account.

Also, for confidentiallity reasons, the banking customer identifier is never the real customer banking account.

And finaly, Postfinance client using ESR don't need to use the banking customer identifier because the banking account used belongs to the client.

**Comment by ravage84** [comment]
@sdespont don't worry, I don't confuse them, though I both mentioned them together. I knew most of what you wrote already. I just wanted to find a pattern within all of this.

> BUT, it is up to the bank to choose how they want to match the customer banking account and 6 digits is a strong limitation if you have a lot of clients (futhermore, the identifier is never reused for another client if the customer close its account).

I'm curious, where do you know these things from?

> It is why a bank can decide to use 7 or 8 digits intead of 6 to match the customer banking account.

Where do you have the statement about "7 or 8 digits" from?

**Comment by sdespont** [comment]
Simply the fatcs. Working with 110 banking accounts, 5 use banking customer idenfier with 7 digits, 1 use 8 digits. All of the banking customer idenfier with 7 or 8 come from 5 differents Raiffeisen.

**Comment by ravage84** [comment]
@sdespont many thanks for this valuable information! Now I understand your problem. ;-)

I think 10 digits as the current uppper limit is OK for now.

**Comment by ravage84** [comment]
@sdespont if you can have a look at [my code review](https://github.com/ravage84/SwissPaymentSlip/pull/27/files/80886c9ecfacdff20ebc6216c979a23e0e8edc8d), we can go forward with your PR. 😎

File: src/OrangePaymentSlipData.php (hunk 3 above)
Comment (sdespont) [reply src/OrangePaymentSlipData.php:69]: I don't agree, if `$this->bankingCustomerId = '';`, `$this->bankingCustomerIdLength = 0;` must be set to `0` and not to `null`.

Futhermore, the specified type is integer, so it should never be null.

File: src/OrangePaymentSlipData.php (hunk 4 above)
Comment (sdespont) [reply src/OrangePaymentSlipData.php:116]: I don't agree, if `$this->bankingCustomerId = '';`, `$this->bankingCustomerIdLength = 0;` must be set to `0` and not to `null`.

Futhermore, the specified type is integer, so it should never be null.

File: src/OrangePaymentSlipData.php (hunk 15 above)
Comment (sdespont) [reply src/OrangePaymentSlipData.php:219]: The goal to set the `bankingCustomerIdLength` is to be sure that the returned value will have this length including zeroes left padding if necessary. If you provide a `bankingCustomerId` equals to `1234` and a `bankingCustomerIdLength` equals to `6`, you expect `001234` otherwise the `bankingCustomerIdLength` has to be `4`.

**Comment by ravage84** [comment]
I would have preferred https://github.com/ravage84/SwissPaymentSlip/pull/27/commits/1037ca07997c46d4c4053d73dc58935c7c7c7e83 to be a separate PR, which we can handle separately.

Also that commit contains seemingly unrelated changes to [composer.lock](https://github.com/ravage84/SwissPaymentSlip/commit/1037ca07997c46d4c4053d73dc58935c7c7c7e83#diff-1da2c7edc898c70e5a79a9997c98cecc).

File: src/PaymentSlipData.php [hunk 23]
```diff
@@ -620,6 +620,13 @@ public function setAmount($amount = 0.0)
         }
         $this->amount = $amount;
 
+        // Fix case where cents are 995 to 999 : the amount must be rounded
+        $francs = intval($amount);
+        $cents = round(($amount - $francs) * 100);
```
Comment (ravage84) [inline src/PaymentSlipData.php]: Round first, split later?

File: composer.lock [hunk 24]
```diff
@@ -60,7 +59,7 @@
                 "constructor",
                 "instantiate"
             ],
-            "time": "2015-06-14 21:17:01"
+            "time": "2015-06-14T21:17:01+00:00"
```
Comment (ravage84) [inline composer.lock:62]: These changes seem unrelated. How come?

File: CHANGELOG.md [hunk 25]
```diff
@@ -9,8 +9,12 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
 
 ### Fixed
+## [0.13.1]
+Fix case where cents are from 995 to 999 - 2017-06-07
```
Comment (ravage84) [inline CHANGELOG.md]: As 0.13.0 is not released yet, it will be all within 0.13.0.
But as I wrote in https://github.com/ravage84/SwissPaymentSlip/pull/27/commits/afea3e618d399fc52629e22a567081374170a2c4#r119148628 please add everythin under "Unreleased" and I'll update everything accodringly when I release a new version/tag.

File: src/OrangePaymentSlipData.php [hunk 26]
```diff
@@ -27,6 +27,9 @@
  */
 class OrangePaymentSlipData extends PaymentSlipData
 {
+    const REFNR_MAX_LENGTH = 26;
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php:31]: I'll add doc blocks for them post-merge.

File: src/OrangePaymentSlipData.php [hunk 27]
```diff
@@ -130,19 +133,19 @@ public function getWithBankingCustomerId()
...
      */
     public function setReferenceNumber($referenceNumber)
     {
         if (!$this->getWithReferenceNumber()) {
             throw new DisabledDataException('reference number');
         }
-                
-        if (strlen($referenceNumber) + $this->bankingCustomerIdLength > 26) {
+
+        if (strlen($referenceNumber) + $this->bankingCustomerIdLength > $this::REFNR_MAX_LENGTH) {
```
Comment (ravage84) [inline src/OrangePaymentSlipData.php]: Why do you use ``this`` instead of ``static``?
http://php.net/manual/en/language.oop5.late-static-bindings.php

File: src/OrangePaymentSlipData.php (hunk 6 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php]: Created separate issue for that https://github.com/ravage84/SwissPaymentSlip/issues/32

File: src/OrangePaymentSlipData.php (hunk 12 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php]: Created separate issue for that https://github.com/ravage84/SwissPaymentSlip/issues/33

File: src/OrangePaymentSlipData.php (hunk 15 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php:219]: Sure, I get that. Just thought about the ``fillZeroes`` parameter. But aparently is the idea that it only applies when the banking customer ID nis not used. https://github.com/ravage84/SwissPaymentSlip/blob/0.12.1/src/OrangePaymentSlipData.php#L210

So 👍

File: src/OrangePaymentSlipData.php (hunk 3 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php:69]: You are right. I was coming from a different angle. Never mind 👍

File: src/OrangePaymentSlipData.php (hunk 4 above)
Comment (ravage84) [reply src/OrangePaymentSlipData.php:116]: You are right. I was coming from a different angle. Never mind 👍

**Review by ravage84 — CHANGES_REQUESTED** [review]
Starts to look good. Some things I'm going to fix post-merge. Please address the few left over.

**Comment by sdespont** [comment]
Marc,

Are you waiting for something else?

Sébastien

**Comment by ravage84** [comment]
@sdespont free time to have a look at it a last time before merging. Sorry, I'm busy s f**k at the moment.

**Comment by ravage84** [comment]
I'm not going to implement or action this, anymore, as this package has become obsolete due to the introduction of the [Swiss QR code payment slip](https://de.wikipedia.org/wiki/QR-Rechnung) and the discontinuation of the old ESR.

One can use https://github.com/sprain/php-swiss-qr-bill for the new Swiss QR  bill.
